[5.0] Implement ChromeDriverCommand

driesvints opened the pull request:
This PR implements the functionality offered by https://github.com/staudenmeir/dusk-updater which was written by @staudenmeir.

This command will allow users to update their chrome driver for their OS. It has a flag that will also allow users to update every binary if they want. I've added this to the install command so the driver is updated whenever the install command is run. This means we don't have to update the docs.

I wrote this as an install command so we can remove the binaries in a future version of Dusk and always have them installed with this command. This will save having them to be packaged with the repo. ~It's also written in such a way that it only installs the binary for the current OS.~

Fixes https://github.com/laravel/dusk/issues/641

driesvints commented:
@staudenmeir I'd greatly appreciate a review from you :)

staudenmeir commented:
Not a review, but things I've noticed:

- The `resources/legacy.json` file is missing.
- Can we reuse the OS detection from `ChromeProcess` (by moving it to a shared place)?
- Since the code is basically a copy of my package, it should/must include some kind of copyright notice.

staudenmeir commented on composer.json:
```diff
@@ -11,6 +11,8 @@
     ],
     "require": {
         "php": ">=7.1.0",
+        "ext-json": "*",
```
I don't think we need to add `ext-json`, it's already required by Laravel.

driesvints replied:
We do since we use `json_` methods in this library. It's always better to explicitly state your dependencies in your composer.json, even if it's used in Laravel.

staudenmeir commented on src/Console/InstallCommand.php:
```diff
@@ -67,6 +57,10 @@ public function handle()
+
+        $this->info('Installing ChromeDriver binary...');
+
+        $this->call('dusk:chrome-driver');
```
We should use the `--all` option here. If you install Laravel & Dusk on a Windows/Mac machine and then run the tests in Homestead, you aren't using the latest ChromeDriver.

driesvints replied:
Hah, I was doubting to do that but you're right. I've updated this.

driesvints commented:
> The resources/legacy.json file is missing.

I've added this as an array in the command.

> Can we reuse the OS detection from ChromeProcess (by moving it to a shared place)?

Yeah I was thinking that as well. That's probably better indeed. I've pushed a class that re-uses this knowledge.

> Since the code is basically a copy of my package, it should/must include some kind of copyright notice.

You're absolutely right. I've updated the license file.

driesvints commented:
@staudenmeir updated.

staudenmeir commented on src/Console/InstallCommand.php:
```diff
@@ -67,6 +57,10 @@ public function handle()
+
+        $this->info('Installing ChromeDriver binaries...');
+
+        $this->call('dusk:chrome-driver', ['--all']);
```
```suggestion
        $this->call('dusk:chrome-driver', ['--all' => true]);
```

driesvints replied:
Fixed. Good catch.

staudenmeir commented on composer.json:
```diff
@@ -11,6 +11,8 @@
     ],
     "require": {
         "php": ">=7.1.0",
+        "ext-json": "*",
```
Now that `legacy.json` is stored in an array, we can actually remove `ext-json`.

driesvints replied:
I noticed some other usages in the library so I'm gonna leave it in.

driesvints commented:
@staudenmeir thanks for all your work on this btw 👍

staudenmeir commented:
When I updated my package, I noticed a potential issue with the new ChromeDriver release strategy:
There will now be beta releases for upcoming Chrome versions and `latestChromeVersion()` might incorrectly select them as the latest version (technically correct, but not really useful).

We can't do anything about that at the moment, we just have to wait for the next release to see what the index page will look like.

taylorotwell commented on LICENSE.md:
```diff
@@ -1,6 +1,7 @@
 The MIT License (MIT)
 
 Copyright (c) Taylor Otwell
+Copyright (c) Jonas Staudenmeir
```
I don't think it's common to add to this file on contributions to open source projects. Would we add everyone that has ever contributed to Dusk?

staudenmeir replied:
We don't have to change the license file, a copyright notice in `ChromeDriverCommand.php` would be sufficient.

driesvints replied:
@staudenmeir I've pushed a commit for this, does this look good to you? https://github.com/laravel/dusk/commit/7e5414b488bbf1e59bbfa77bd0a284cc228678f8

staudenmeir replied:
Please also include the package URL.

driesvints replied:
Done!

staudenmeir replied:
Thanks!

driesvints commented:
@staudenmeir I've pushed a commit for this, does this look good to you? https://github.com/laravel/dusk/commit/7e5414b488bbf1e59bbfa77bd0a284cc228678f8

bobbybouwmann commented:
@driesvints I did found some errors when running this command
```
Facebook\WebDriver\Exception\SessionNotCreatedException: session not created: Chrome version must be between 70 and 73
```
So for me it installed the latest Chrome which is 74, as you also can see on the homepage: http://chromedriver.chromium.org/home

However the Facebook driver doesn't support this version just yet, at least that's what the error is telling me. I've been looking into the `facebook\php-webdriver` package but couldn't find where this check is done!

A work around for now was downloading an older version manually, but the command currently doesn't work whenever a new version is released and not updated in the `php-webdriver` package.

driesvints commented:
Hey @bobbybouwmann, did you read through https://github.com/laravel/dusk/issues/641 to see if any of the answers helps you?

> So for me it installed the latest Chrome which is 74, as you also can see on the homepage

The error says `Chrome version must be between 70 and 73` so did you try to install version 71 or 72? What's your current chrome browser version? Did you try to upgrade your Chrome browser and quit and restart afterwards? Perhaps even reboot your computer to make sure no processes are lingering anymore?

> However the Facebook driver doesn't support this version just yet

I guess I can't help with this and you'll need to use a lower version for now...

staudenmeir commented:
> However the Facebook driver doesn't support this version just yet, at least that's what the error is telling me. 

Do you mean the error that you posted? This error is not caused by an incompatibility with the `facebook/webdriver` package.

bobbybouwmann commented:
@driesvints I had indeed 74 installed, but that was because of the `php artisan du`sk:chrome-driver` command. I installed 72 and everything started working again! Just wanted to give a heads up!
